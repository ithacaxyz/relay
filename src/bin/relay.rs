//! # Ithaca Relay
//!
//! A relay service that sponsors transactions for EIP-7702 accounts.

use clap::Parser;
use relay::{
    cli::*,
    otlp::{OtelConfig, OtelGuard},
};
use tracing::debug;
use tracing_subscriber::prelude::*;

fn init_tracing_subscriber() -> Option<OtelGuard> {
    let registry = tracing_subscriber::registry().with(
        tracing_subscriber::fmt::layer()
            .with_filter(tracing_subscriber::filter::EnvFilter::from_default_env()),
    );

    if let Some(cfg) = OtelConfig::load() {
        let guard = cfg.provider();
        registry.with(guard.layer()).init();
        Some(guard)
    } else {
        registry.init();
        None
    }
}

#[tokio::main]
async fn main() {
    // Enable backtraces unless a RUST_BACKTRACE value has already been explicitly provided.
    if std::env::var_os("RUST_BACKTRACE").is_none() {
        unsafe { std::env::set_var("RUST_BACKTRACE", "1") };
    }

    let _guard = init_tracing_subscriber();
    if _guard.is_some() {
        debug!("opentelemetry initialized");
    }

    let args = Args::parse();
    if let Err(err) = args.run().await {
        eprintln!("Error: {err:?}");
        std::process::exit(1);
    }
}
