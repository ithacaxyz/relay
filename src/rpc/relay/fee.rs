use crate::{
    chains::Chain,
    error::{KeysError, QuoteError, RelayError},
    provider::ProviderExt,
    rpc::relay::manager::Manager,
    types::{
        Account, AssetDiffs, GasEstimate, Intent, IntentKind, Key, KeyWith712Signer, Orchestrator,
        OrchestratorContract, PartialAction, Quote, Signature, SignedCalls, Transfer,
    },
};
use alloy::{
    consensus::{SignableTransaction, TxEip1559},
    eips::eip7702::constants::EIP7702_DELEGATION_DESIGNATOR,
    primitives::{Address, Bytes, U256, bytes},
    providers::{
        Provider,
        utils::{EIP1559_FEE_ESTIMATION_PAST_BLOCKS, Eip1559Estimator},
    },
    rpc::types::state::{AccountOverride, StateOverridesBuilder},
    sol_types::{SolCall, SolValue},
};
use futures_util::{TryFutureExt, future::try_join4};
use tracing::{debug, instrument};

#[async_trait::async_trait]
pub trait FeeEstimator: Manager {
    /// Estimate fees for executing actions
    #[instrument(skip_all)]
    async fn estimate_fee(
        &self,
        request: PartialAction,
        token: Address,
        authorization_address: Option<Address>,
        account_key: Key,
        key_slot_override: bool,
        intent_kind: IntentKind,
    ) -> Result<(AssetDiffs, Quote), RelayError> {
        let chain = self
            .chains()
            .get(request.chain_id)
            .ok_or(RelayError::UnsupportedChain(request.chain_id))?;

        let provider = chain.provider.clone();
        let Some(token) = self.fee_tokens().find(request.chain_id, &token) else {
            return Err(QuoteError::UnsupportedFeeToken(token).into());
        };

        // create key
        let mock_key = KeyWith712Signer::random_admin(account_key.keyType)
            .map_err(RelayError::from)
            .and_then(|k| k.ok_or_else(|| RelayError::Keys(KeysError::UnsupportedKeyType)))?;

        // mocking key storage for the eoa, and the balance for the mock signer
        let overrides = StateOverridesBuilder::with_capacity(2)
            // simulateV1Logs requires it, so the function can only be called under a testing
            // environment
            .append(self.simulator(), AccountOverride::default().with_balance(U256::MAX))
            .append(self.orchestrator(), AccountOverride::default().with_balance(U256::MAX))
            .append(
                request.intent.eoa,
                AccountOverride::default()
                    .with_balance(U256::MAX.div_ceil(2.try_into().unwrap()))
                    .with_state_diff(if key_slot_override {
                        account_key.storage_slots()
                    } else {
                        Default::default()
                    })
                    // we manually etch the 7702 designator since we do not have a signed auth item
                    .with_code_opt(authorization_address.map(|addr| {
                        Bytes::from([&EIP7702_DELEGATION_DESIGNATOR, addr.as_slice()].concat())
                    })),
            )
            .build();

        let account = Account::new(request.intent.eoa, &provider).with_overrides(overrides.clone());

        let (orchestrator, delegation, fee_history, eth_price) = try_join4(
            // fetch orchestrator from the account and ensure it is supported
            async {
                let orchestrator = account.get_orchestrator().await?;
                if !self.is_supported_orchestrator(&orchestrator) {
                    return Err(RelayError::UnsupportedOrchestrator(orchestrator));
                }
                Ok(Orchestrator::new(orchestrator, &provider).with_overrides(overrides.clone()))
            },
            // fetch delegation from the account and ensure it is supported
            self.has_supported_delegation(&account).map_err(RelayError::from),
            // fetch chain fees
            provider
                .get_fee_history(
                    EIP1559_FEE_ESTIMATION_PAST_BLOCKS,
                    Default::default(),
                    &[self.priority_fee_percentile()],
                )
                .map_err(RelayError::from),
            // fetch price in eth
            async {
                // TODO: only handles eth as native fee token
                Ok(self.price_oracle().eth_price(token.kind).await)
            },
        )
        .await?;

        let native_fee_estimate = Eip1559Estimator::default().estimate(
            fee_history.latest_block_base_fee().unwrap_or_default(),
            &fee_history.reward.unwrap_or_default(),
        );

        let Some(eth_price) = eth_price else {
            return Err(QuoteError::UnavailablePrice(token.address).into());
        };
        let payment_per_gas = if intent_kind.is_single() {
            (native_fee_estimate.max_fee_per_gas as f64 * 10u128.pow(token.decimals as u32) as f64)
                / f64::from(eth_price)
        } else {
            // todo: is_multi_input should take a fee as well eventually
            0f64
        };

        // fill intent
        let mut intent = Intent {
            eoa: request.intent.eoa,
            executionData: request.intent.execution_data.clone(),
            nonce: request.intent.nonce,
            payer: request.intent.payer.unwrap_or_default(),
            paymentToken: token.address,
            paymentRecipient: self.fee_recipient(),
            supportedAccountImplementation: delegation,
            encodedPreCalls: request
                .intent
                .pre_calls
                .into_iter()
                .map(|pre_call| pre_call.abi_encode().into())
                .collect(),
            encodedFundTransfers: request
                .intent
                .fund_transfers
                .into_iter()
                .map(|(token, amount)| Transfer { token, amount }.abi_encode().into())
                .collect(),
            ..Default::default()
        };

        let extra_payment = self.estimate_extra_fee(&chain, &intent).await?
            * U256::from(10u128.pow(token.decimals as u32))
            / eth_price;

        let intrinsic_gas = approx_intrinsic_cost(
            &OrchestratorContract::executeCall {
                isMultiChain: !intent_kind.is_single(),
                encodedIntent: intent.abi_encode().into(),
            }
            .abi_encode(),
            authorization_address.is_some(),
        );

        let initial_payment = U256::from(intrinsic_gas as f64 * payment_per_gas) + extra_payment;

        intent.set_legacy_payment_amount(initial_payment);

        // sign intent
        let signature = mock_key
            .sign_typed_data(
                &intent.as_eip712().map_err(RelayError::from)?,
                &orchestrator
                    .eip712_domain(intent.is_multichain())
                    .await
                    .map_err(RelayError::from)?,
            )
            .await
            .map_err(RelayError::from)?;

        intent.signature = Signature {
            innerSignature: signature,
            keyHash: account_key.key_hash(),
            prehash: request.prehash,
        }
        .abi_encode_packed()
        .into();

        if !intent.encodedFundTransfers.is_empty() {
            // todo: the contract version is broken, and any signature will pass.
            intent.funder = self.contracts().funder.address;
        }

        // todo: simulate with executeMultiChain if intent.is_multichain
        // we estimate gas and fees
        let (asset_diff, sim_result) = orchestrator
            .simulate_execute(
                self.simulator(),
                &intent,
                account_key.keyType,
                payment_per_gas,
                self.asset_info().clone(),
            )
            .await?;

        // todo: re-evaluate if this is still necessary
        let gas_estimate = GasEstimate::from_combined_gas(
            sim_result.gCombined.to(),
            intrinsic_gas,
            self.quote_config(),
        );

        debug!(eoa = %request.intent.eoa, gas_estimate = ?gas_estimate, "Estimated intent");

        // Fill combinedGas and empty dummy signature
        intent.combinedGas = U256::from(gas_estimate.intent);
        intent.signature = bytes!("");
        intent.funderSignature = bytes!("");

        // Calculate amount with updated paymentPerGas
        if !intent_kind.is_single() {
            // todo: temporary
            intent.set_legacy_payment_amount(U256::ZERO);
        } else {
            intent.set_legacy_payment_amount(
                extra_payment + U256::from((payment_per_gas * gas_estimate.tx as f64).ceil()),
            )
        }

        let quote = Quote {
            chain_id: request.chain_id,
            payment_token_decimals: token.decimals,
            output: intent,
            extra_payment,
            eth_price,
            tx_gas: gas_estimate.tx,
            native_fee_estimate,
            authorization_address,
            orchestrator: *orchestrator.address(),
            is_multi_chain: !intent_kind.is_single(),
        };

        Ok((asset_diff, quote))
    }

    /// Estimates additional fees to be paid for a intent (e.g L1 DA fees).
    ///
    /// Returns fees in ETH.
    #[instrument(skip_all)]
    async fn estimate_extra_fee(&self, chain: &Chain, intent: &Intent) -> Result<U256, RelayError> {
        // Include the L1 DA fees if we're on an OP rollup.
        let fee = if chain.is_optimism {
            // Create a dummy transactions with all fields set to max values to make sure that
            // calldata is largest possible
            let tx = TxEip1559 {
                chain_id: chain.chain_id,
                nonce: u64::MAX,
                gas_limit: u64::MAX,
                max_fee_per_gas: u128::MAX,
                max_priority_fee_per_gas: u128::MAX,
                to: (!Address::ZERO).into(),
                input: intent.encode_execute(false),
                ..Default::default()
            };
            let signature = alloy::signers::Signature::new(U256::MAX, U256::MAX, true);

            let encoded = {
                let tx = tx.into_signed(signature);
                let mut buf = Vec::with_capacity(tx.eip2718_encoded_length());
                tx.eip2718_encode(&mut buf);
                buf
            };

            chain.provider.estimate_l1_fee(encoded.into()).await?
        } else {
            U256::ZERO
        };

        Ok(fee)
    }
}

/// Approximates the intrinsic cost of a transaction.
///
/// This function assumes Prague rules.
fn approx_intrinsic_cost(input: &[u8], has_auth: bool) -> u64 {
    use alloy::eips::eip7702::constants::PER_EMPTY_ACCOUNT_COST;

    let zero_data_len = input.iter().filter(|v| **v == 0).count() as u64;
    let non_zero_data_len = input.len() as u64 - zero_data_len;
    let non_zero_data_multiplier = 4; // as defined in istanbul
    let standard_token_cost = 4;
    let tokens = zero_data_len + non_zero_data_len * non_zero_data_multiplier;

    // for 7702 designations there is an additional gas charge
    //
    // note: this is not entirely accurate, as there is also a gas refund in 7702, but at this
    // point it is not possible to compute the gas refund, so it is an overestimate, as we also
    // need to charge for the account being presumed empty.
    let auth_cost = if has_auth { PER_EMPTY_ACCOUNT_COST } else { 0 };

    21000 + auth_cost + tokens * standard_token_cost
}
