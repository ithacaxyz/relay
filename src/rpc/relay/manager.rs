use crate::{
    asset::AssetInfoServiceHandle,
    chains::Chains,
    config::QuoteConfig,
    error::{AuthError, RelayError},
    price::PriceOracle,
    storage::StorageApi,
    types::{Account, FeeTokens, VersionedContracts},
};
use alloy::{
    eips::eip7702::constants::EIP7702_DELEGATION_DESIGNATOR,
    primitives::{Address, Bytes, ChainId},
    providers::{DynProvider, Provider},
    rpc::types::state::StateOverridesBuilder,
};
use tracing::instrument;

#[async_trait::async_trait]
pub trait Manager {
    /// Get chains
    fn chains(&self) -> &Chains;

    /// Get contracts
    fn contracts(&self) -> &VersionedContracts;

    /// Get orchestrator address
    fn orchestrator(&self) -> Address {
        self.contracts().orchestrator.address
    }

    /// Previously deployed orchestrators.
    fn legacy_orchestrators(&self) -> impl Iterator<Item = Address> {
        self.contracts().legacy_orchestrators.iter().map(|c| c.address)
    }

    /// Get delegation implementation address
    fn delegation_implementation(&self) -> Address {
        self.contracts().delegation_implementation.address
    }

    /// Previously deployed delegation implementations.
    fn legacy_delegations(&self) -> impl Iterator<Item = Address> {
        self.contracts().legacy_delegations.iter().map(|c| c.address)
    }

    /// The simulator address.
    fn simulator(&self) -> Address {
        self.contracts().simulator.address
    }

    /// Get fee recipient
    fn fee_recipient(&self) -> Address;

    /// Get priority fee percentile
    fn priority_fee_percentile(&self) -> f64;

    /// Get quote config
    fn quote_config(&self) -> &QuoteConfig;

    /// Get fee tokens
    fn fee_tokens(&self) -> &FeeTokens;

    /// Get price oracle
    fn price_oracle(&self) -> &PriceOracle;

    /// Get asset info
    fn asset_info(&self) -> &AssetInfoServiceHandle;

    /// Get storage
    fn storage(&self) -> &dyn StorageApi;

    /// Get a provider for the given chain ID
    fn provider(&self, chain_id: ChainId) -> Result<DynProvider, RelayError> {
        self.chains()
            .get(chain_id)
            .map(|chain| chain.provider.clone())
            .ok_or(RelayError::UnsupportedChain(chain_id))
    }

    /// Checks if the orchestrator is supported.
    fn is_supported_orchestrator(&self, orchestrator: &Address) -> bool {
        self.orchestrator() == *orchestrator
            || self.legacy_orchestrators().any(|c| c == *orchestrator)
    }

    /// Checks if the account has a supported delegation implementation. If so, returns it.
    async fn has_supported_delegation<P: Provider + Clone>(
        &self,
        account: &Account<P>,
    ) -> Result<Address, RelayError> {
        let address = self.get_delegation_implementation(account).await?;
        if self.delegation_implementation() == address
            || self.legacy_delegations().any(|c| c == address)
        {
            return Ok(address);
        }
        Err(AuthError::InvalidDelegation(address).into())
    }

    /// Ensures the account has the latest delegation implementation. Otherwise, returns error.
    async fn ensure_latest_delegation<P: Provider + Clone>(
        &self,
        account: &Account<P>,
    ) -> Result<(), RelayError> {
        let address = self.has_supported_delegation(account).await?;
        if self.delegation_implementation() != address {
            return Err(AuthError::InvalidDelegation(address).into());
        }
        Ok(())
    }

    /// Returns the delegation implementation address from the requested account.
    ///
    /// It will return error if the delegation proxy is invalid.
    #[instrument(skip_all)]
    async fn get_delegation_implementation<P: Provider + Clone>(
        &self,
        account: &Account<P>,
    ) -> Result<Address, RelayError> {
        if let Some(delegation) = account.delegation_implementation().await? {
            return Ok(delegation);
        }

        // Attempt to retrieve the delegation proxy from storage, since it might not be
        // deployed yet.
        let Some(stored) = self.storage().read_account(&account.address()).await? else {
            return Err(RelayError::Auth(AuthError::EoaNotDelegated(account.address()).boxed()));
        };

        let address = account.address();
        let account = account.clone().with_overrides(
            StateOverridesBuilder::default()
                .with_code(
                    address,
                    Bytes::from(
                        [
                            &EIP7702_DELEGATION_DESIGNATOR,
                            stored.signed_authorization.address().as_slice(),
                        ]
                        .concat(),
                    ),
                )
                .build(),
        );

        account.delegation_implementation().await?.ok_or_else(|| {
            RelayError::Auth(
                AuthError::InvalidDelegationProxy(*stored.signed_authorization.address()).boxed(),
            )
        })
    }
}
