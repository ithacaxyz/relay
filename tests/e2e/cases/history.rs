use crate::e2e::{
    environment::mint_erc20s,
    eoa::{MockAccount, MockAccountBuilder},
    *,
};
use alloy::primitives::{Address, U256};
use chrono::Utc;
use eyre::Result;
use relay::{
    rpc::{RelayApiClient, adjust_balance_for_decimals},
    storage::StorageApi,
    types::{
        Call, DiffDirection, HistoricalPrice, IERC20,
        rpc::{BundleId, CallHistoryEntry, GetCallsHistoryParameters, SortDirection},
    },
};
use std::time::Duration;

#[tokio::test(flavor = "multi_thread")]
async fn test_multichain_calls_history_mixed_bundles() -> Result<()> {
    let env = Environment::setup_multi_chain(2).await?;

    // Create fresh account without ERC20 minting (includes initialization bundle)
    let account = MockAccountBuilder::new().no_erc20_mint().build(&env).await?;

    // Manually mint ERC20 on chain0 and chain1
    mint_erc20s(&[env.erc20], &[account.address], env.provider_for(0)).await?;
    mint_erc20s(&[env.erc20], &[account.address], env.provider_for(1)).await?;

    // Bundle 1: Single-chain on chain0
    let bundle1 = send_bundle(&env, &account, 0, U256::from(1), env.erc20).await?;

    // Sleep to ensure different timestamps for price testing
    tokio::time::sleep(Duration::from_secs(1)).await;

    // Bundle 2: Single-chain on chain1
    let bundle2 = send_bundle(&env, &account, 1, U256::from(1), env.erc20).await?;

    // Sleep to ensure different timestamps for price testing
    tokio::time::sleep(Duration::from_secs(1)).await;

    // Bundle 3: Interop - move more funds than available on chain1
    let bundle3 = {
        // Get decimals for both chains
        let decimals_0 = IERC20::new(env.erc20, env.provider_for(0)).decimals().call().await?;
        let decimals_1 = IERC20::new(env.erc20, env.provider_for(1)).decimals().call().await?;

        // Get balances on both chains
        let balance_chain0 =
            IERC20::new(env.erc20, env.provider_for(0)).balanceOf(account.address).call().await?;
        let balance_chain1 =
            IERC20::new(env.erc20, env.provider_for(1)).balanceOf(account.address).call().await?;

        // Adjust balance from chain0 to chain1 decimals
        let balance_chain0_adjusted =
            adjust_balance_for_decimals(balance_chain0, decimals_0, decimals_1);

        // Try to send on chain1: chain1_balance + half of chain0_balance (requires cross-chain)
        let amount_to_send = balance_chain1 + (balance_chain0_adjusted / U256::from(2));

        send_bundle(&env, &account, 1, amount_to_send, env.erc20).await?
    };

    populate_historical_prices(&env, &[bundle1, bundle2, bundle3]).await?;

    // Test 1: Fetch all history (ascending order)
    let history = get_history(&env, account.address, None, 10, SortDirection::Asc).await?;
    assert_asset_diffs_populated(&history);
    assert_eq!(history.len(), 4, "Expected 4 bundles in history");

    let bundle_ids: Vec<_> = history.iter().map(|e| e.id).collect();
    assert!(bundle_ids.contains(&bundle1));
    assert!(bundle_ids.contains(&bundle2));
    assert!(bundle_ids.contains(&bundle3));

    // Verify the first entry (initialization) is not one of our sent bundles
    assert_ne!(history[0].id, bundle1, "Init bundle should not be bundle1");
    assert_ne!(history[0].id, bundle2, "Init bundle should not be bundle2");
    assert_ne!(history[0].id, bundle3, "Init bundle should not be bundle3");

    // Verify bundle3 (interop) has quotes from both src and dst transactions
    let bundle3_entry = history.iter().find(|e| e.id == bundle3).expect("bundle3 should exist");
    assert_eq!(
        bundle3_entry.capabilities.quotes.len(),
        2,
        "Interop bundle should have 2 quotes (1 from src_txs, 1 from dst_txs)"
    );

    // Test 2: Fetch history in descending order
    let history_desc = get_history(&env, account.address, None, 10, SortDirection::Desc).await?;
    assert_asset_diffs_populated(&history);

    assert_eq!(history_desc.len(), 4);
    // Verify it's in reverse order
    assert_eq!(history_desc[0].id, bundle3, "Most recent should be bundle3");
    assert_eq!(history_desc[3].id, history[0].id, "Oldest should match first in asc");

    // Test 3: Limit to 2 bundles
    let history_limited = get_history(&env, account.address, None, 2, SortDirection::Asc).await?;
    assert_asset_diffs_populated(&history);

    assert_eq!(history_limited.len(), 2);
    assert_eq!(history_limited[0].id, history[0].id);
    assert_eq!(history_limited[1].id, history[1].id);

    // Test 4: Use index to paginate (skip first 2, get next 2)
    let history_paginated =
        get_history(&env, account.address, Some(2), 2, SortDirection::Asc).await?;
    assert_asset_diffs_populated(&history);

    assert_eq!(history_paginated.len(), 2);
    assert_eq!(history_paginated[0].id, history[2].id);
    assert_eq!(history_paginated[1].id, history[3].id);

    // Test 5: Index beyond available items
    let history_beyond =
        get_history(&env, account.address, Some(10), 10, SortDirection::Asc).await?;
    assert_eq!(history_beyond.len(), 0);

    // Test 6: Verify timestamps are increasing (ascending order)
    for i in 0..history.len() - 1 {
        assert!(
            history[i].timestamp <= history[i + 1].timestamp,
            "Timestamps should be in ascending order"
        );
    }

    Ok(())
}

async fn send_bundle(
    env: &Environment,
    account: &MockAccount,
    chain_idx: usize,
    amount: U256,
    fee_token: Address,
) -> Result<BundleId> {
    let response = env
        .relay_endpoint
        .prepare_calls(PrepareCallsParameters {
            calls: vec![Call::transfer(env.erc20, Address::random(), amount)],
            chain_id: env.chain_id_for(chain_idx),
            from: Some(account.address),
            capabilities: PrepareCallsCapabilities {
                meta: Meta { fee_token: Some(fee_token), ..Default::default() },
                ..Default::default()
            },
            key: Some(account.key.to_call_key()),
            ..Default::default()
        })
        .await?;

    let is_multichain = response.context.quote().unwrap().ty().multi_chain_root.is_some();
    let signature = account.key.sign_payload_hash(response.digest).await?;
    let bundle_id = send_prepared_calls(env, &account.key, signature, response.context).await?;

    let status = await_calls_status(env, bundle_id).await?;
    assert!(status.status.is_confirmed());
    if is_multichain {
        assert!(status.capabilities.unwrap().interop_status.unwrap().is_done());
    }

    Ok(bundle_id)
}

/// Helper to fetch call history with given parameters
async fn get_history(
    env: &Environment,
    address: Address,
    index: Option<u64>,
    limit: u64,
    sort: SortDirection,
) -> Result<Vec<CallHistoryEntry>> {
    env.relay_endpoint
        .get_calls_history(GetCallsHistoryParameters { address, index, limit, sort })
        .await
        .map_err(Into::into)
}

/// Helper to verify that asset diffs and fee totals are populated
fn assert_asset_diffs_populated(entries: &[CallHistoryEntry]) {
    for entry in entries {
        let asset_diff = &entry.capabilities.asset_diff;

        assert!(!asset_diff.fee_totals.is_empty());
        assert!(!asset_diff.asset_diffs.is_empty());

        for chain_diffs in asset_diff.asset_diffs.values() {
            for (_, diffs) in &chain_diffs.0 {
                for diff in diffs {
                    assert!(diff.fiat.is_some(),);
                }
            }
        }
    }
}

/// Test that verifies exact asset diff values match on-chain execution with the correct payer fee
/// removed (only the transfer amount should show up).
#[tokio::test(flavor = "multi_thread")]
async fn test_calls_history_exact_asset_diffs() -> Result<()> {
    let env = Environment::setup().await?;
    let account = MockAccountBuilder::new().build(&env).await?;
    let transfer_amount = U256::from(1000);
    let bundle_id = send_bundle(&env, &account, 0, transfer_amount, env.erc20).await?;
    populate_historical_prices(&env, &[bundle_id]).await?;

    let history = get_history(&env, account.address, None, 10, SortDirection::Asc).await?;
    let entry = history.iter().find(|e| e.id == bundle_id).unwrap();

    let quote = &entry.capabilities.quotes[0];
    let eoa_diffs = entry
        .capabilities
        .asset_diff
        .asset_diffs
        .get(&quote.chain_id)
        .unwrap()
        .0
        .iter()
        .find(|(addr, _)| addr == &account.address)
        .map(|(_, diffs)| diffs)
        .unwrap();

    let erc20_diff =
        eoa_diffs.iter().find(|d| d.address == Some(env.erc20)).expect("ERC20 diff should exist");

    assert_eq!(erc20_diff.value, transfer_amount,); // fee got removed correctly
    assert_eq!(erc20_diff.direction, DiffDirection::Outgoing,);
    assert!(!erc20_diff.recipients.is_empty(),);
    assert!(erc20_diff.fiat.is_some());

    let fee_total_price = entry
        .capabilities
        .asset_diff
        .fee_totals
        .get(&quote.chain_id)
        .expect("Fee total should exist for chain");

    assert!(fee_total_price.value > 0.0,);
    assert_eq!(fee_total_price.currency, "usd",);

    Ok(())
}

/// Helper to populate historical USD prices for bundles
async fn populate_historical_prices(env: &Environment, _bundle_ids: &[BundleId]) -> Result<()> {
    let now = Utc::now().timestamp() as u64;
    let (erc20_asset_uid, _) =
        env.relay_handle.chains.asset(env.chain_id(), env.erc20).expect("erc20 asset should exist");

    let mut prices = Vec::new();

    let (native_asset_uid, _) = env
        .relay_handle
        .chains
        .asset(env.chain_id(), Address::ZERO)
        .expect("native asset should exist");

    for asset_uid in [erc20_asset_uid, native_asset_uid] {
        prices.push(HistoricalPrice {
            asset_uid: asset_uid.clone(),
            timestamp: now - 60,
            usd_price: 1.0,
        });
        prices.push(HistoricalPrice {
            asset_uid: asset_uid.clone(),
            timestamp: now,
            usd_price: 2.0,
        });
        prices.push(HistoricalPrice {
            asset_uid: asset_uid.clone(),
            timestamp: now + 60,
            usd_price: 3.0,
        });
    }

    env.relay_handle.storage.store_historical_usd_prices(prices).await?;

    Ok(())
}
