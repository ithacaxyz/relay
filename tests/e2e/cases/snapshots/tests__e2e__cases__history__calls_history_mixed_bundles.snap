---
source: tests/e2e/cases/history.rs
expression: history
---
[
  {
    "id": "[bundle_id]",
    "index": 0,
    "status": 200,
    "timestamp": "[timestamp]",
    "transactions": [
      {
        "chainId": "0x7a69",
        "transactionHash": "[hash]"
      }
    ],
    "key": "[key]",
    "capabilities": {
      "feeTotals": {},
      "assetDiffs": {},
      "quotes": [
        {
          "chainId": "0x7a69",
          "intent": {
            "eoa": "[address]",
            "executionData": "[execution_data]",
            "nonce": "[nonce]",
            "payer": "0x0000000000000000000000000000000000000000",
            "paymentToken": "0x0000000000000000000000000000000000000000",
            "paymentMaxAmount": "[amount]",
            "combinedGas": "0x31a6a",
            "encodedPreCalls": "[precalls]",
            "encodedFundTransfers": "[fund_transfers]",
            "settler": "0x0000000000000000000000000000000000000000",
            "expiry": "0x0",
            "isMultichain": false,
            "funder": "0x0000000000000000000000000000000000000000",
            "funderSignature": "[signature]",
            "settlerContext": "0x",
            "paymentAmount": "[amount]",
            "paymentRecipient": "0x0000000000000000000000000000000000000000",
            "signature": "[signature]",
            "paymentSignature": "[signature]",
            "supportedAccountImplementation": "0xb19b36b1456e65e3a6d514d3f715f204bd59f431"
          },
          "extraPayment": "[payment]",
          "ethPrice": "0xde0b6b3a7640000",
          "paymentTokenDecimals": 18,
          "txGas": "0x63526",
          "nativeFeeEstimate": {
            "maxFeePerGas": "[fee]",
            "maxPriorityFeePerGas": "0x1"
          },
          "authorizationAddress": "0x29a79095352a718b3d7fe84e1f14e9f34a35598e",
          "additionalAuthorization": null,
          "orchestrator": "0x700b6a60ce7eaaea56f065753d8dcb9653dbad35",
          "feeTokenDeficit": "0x0"
        }
      ]
    }
  },
  {
    "id": "[bundle_id]",
    "index": 1,
    "status": 200,
    "timestamp": "[timestamp]",
    "transactions": [
      {
        "chainId": "0x7a69",
        "transactionHash": "[hash]"
      }
    ],
    "key": "[key]",
    "capabilities": {
      "feeTotals": {},
      "assetDiffs": {},
      "quotes": [
        {
          "chainId": "0x7a69",
          "intent": {
            "eoa": "[address]",
            "executionData": "[execution_data]",
            "nonce": "[nonce]",
            "payer": "0x0000000000000000000000000000000000000000",
            "paymentToken": "0xed1db453c3156ff3155a97ad217b3087d5dc5f6e",
            "paymentMaxAmount": "[amount]",
            "combinedGas": "0x24419",
            "encodedPreCalls": "[precalls]",
            "encodedFundTransfers": "[fund_transfers]",
            "settler": "0x0000000000000000000000000000000000000000",
            "expiry": "0x0",
            "isMultichain": false,
            "funder": "0x0000000000000000000000000000000000000000",
            "funderSignature": "[signature]",
            "settlerContext": "0x",
            "paymentAmount": "[amount]",
            "paymentRecipient": "0x0000000000000000000000000000000000000000",
            "signature": "[signature]",
            "paymentSignature": "[signature]",
            "supportedAccountImplementation": "0xb19b36b1456e65e3a6d514d3f715f204bd59f431"
          },
          "extraPayment": "[payment]",
          "ethPrice": "0xde0b6b3a7640000",
          "paymentTokenDecimals": 18,
          "txGas": "0x4d3c6",
          "nativeFeeEstimate": {
            "maxFeePerGas": "[fee]",
            "maxPriorityFeePerGas": "0x1"
          },
          "authorizationAddress": null,
          "additionalAuthorization": null,
          "orchestrator": "0x700b6a60ce7eaaea56f065753d8dcb9653dbad35",
          "feeTokenDeficit": "0x0"
        }
      ]
    }
  },
  {
    "id": "[bundle_id]",
    "index": 2,
    "status": 200,
    "timestamp": "[timestamp]",
    "transactions": [
      {
        "chainId": "0x7a6a",
        "transactionHash": "[hash]"
      }
    ],
    "key": "[key]",
    "capabilities": {
      "feeTotals": {},
      "assetDiffs": {},
      "quotes": [
        {
          "chainId": "0x7a6a",
          "intent": {
            "eoa": "[address]",
            "executionData": "[execution_data]",
            "nonce": "[nonce]",
            "payer": "0x0000000000000000000000000000000000000000",
            "paymentToken": "0xed1db453c3156ff3155a97ad217b3087d5dc5f6e",
            "paymentMaxAmount": "[amount]",
            "combinedGas": "0x3f254",
            "encodedPreCalls": "[precalls]",
            "encodedFundTransfers": "[fund_transfers]",
            "settler": "0x0000000000000000000000000000000000000000",
            "expiry": "0x0",
            "isMultichain": false,
            "funder": "0x0000000000000000000000000000000000000000",
            "funderSignature": "[signature]",
            "settlerContext": "0x",
            "paymentAmount": "[amount]",
            "paymentRecipient": "0x0000000000000000000000000000000000000000",
            "signature": "[signature]",
            "paymentSignature": "[signature]",
            "supportedAccountImplementation": "0xb19b36b1456e65e3a6d514d3f715f204bd59f431"
          },
          "extraPayment": "[payment]",
          "ethPrice": "0xde0b6b3a7640000",
          "paymentTokenDecimals": 6,
          "txGas": "0x7207e",
          "nativeFeeEstimate": {
            "maxFeePerGas": "[fee]",
            "maxPriorityFeePerGas": "0x1"
          },
          "authorizationAddress": "0x29a79095352a718b3d7fe84e1f14e9f34a35598e",
          "additionalAuthorization": null,
          "orchestrator": "0x700b6a60ce7eaaea56f065753d8dcb9653dbad35",
          "feeTokenDeficit": "0x0"
        }
      ]
    }
  },
  {
    "id": "[bundle_id]",
    "index": 3,
    "status": 200,
    "timestamp": "[timestamp]",
    "transactions": [
      {
        "chainId": "0x7a69",
        "transactionHash": "[hash]"
      },
      {
        "chainId": "0x7a6a",
        "transactionHash": "[hash]"
      }
    ],
    "key": "[key]",
    "capabilities": {
      "feeTotals": {},
      "assetDiffs": {},
      "quotes": [
        {
          "chainId": "0x7a6a",
          "intent": {
            "eoa": "[address]",
            "executionData": "[execution_data]",
            "nonce": "[nonce]",
            "payer": "0x0000000000000000000000000000000000000000",
            "paymentToken": "0xed1db453c3156ff3155a97ad217b3087d5dc5f6e",
            "paymentMaxAmount": "[amount]",
            "combinedGas": "0x378d9",
            "encodedPreCalls": "[precalls]",
            "encodedFundTransfers": "[fund_transfers]",
            "settler": "0x0c8e79f3534b00d9a3d4a856b665bf4ebc22f2ba",
            "expiry": "0x0",
            "isMultichain": true,
            "funder": "0xa15bb66138824a1c7167f5e85b957d04dd34e468",
            "funderSignature": "[signature]",
            "settlerContext": "0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000007a69",
            "paymentAmount": "[amount]",
            "paymentRecipient": "0x0000000000000000000000000000000000000000",
            "signature": "[signature]",
            "paymentSignature": "[signature]",
            "supportedAccountImplementation": "0xb19b36b1456e65e3a6d514d3f715f204bd59f431"
          },
          "extraPayment": "[payment]",
          "ethPrice": "0xde0b6b3a7640000",
          "paymentTokenDecimals": 6,
          "txGas": "0x6276d",
          "nativeFeeEstimate": {
            "maxFeePerGas": "[fee]",
            "maxPriorityFeePerGas": "0x1"
          },
          "authorizationAddress": null,
          "additionalAuthorization": null,
          "orchestrator": "0x700b6a60ce7eaaea56f065753d8dcb9653dbad35",
          "feeTokenDeficit": "0x0"
        }
      ]
    }
  }
]
